Add rudimentary NestedTensor.sum(dim)

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82387

A first step towards adding dimension-wise reductions to NestedTensor, 
- Assumes tensors in the nested tensor as well as the buffer of the nested tensor are contiguous
- Always enforces `keepdim=True`
- Only supports reduction across the last dimension
- No support for acctype (`dtype` argument)
- No autograd support
- CPU only

Next steps would be to add support for the above. For now this basic support is for prototyping to make sure `NestedTensor` can be used as an API for segment reductions.

mikaylagawarecki committed:
Add basic Nested Tensor.sum(dim)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic Nested Tensor.sum(dim)"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add rudimentary NestedTensor.sum(dim)"




[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4642,6 +4642,8 @@
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    NestedTensorCPU: NestedTensor_sum__dim
```
very tiny naming nit: why two underscores between `sum` and `dim`? I read this as `sum_` followed by `dim` indicating an in-place op

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -665,6 +665,80 @@ Tensor& NestedTensor_mul__Scalar(Tensor& self, const Scalar& other) {
+
+  auto nt_input = get_nested_tensor_impl(self);
+  TORCH_CHECK(
+      nested_tensor_impl_is_contiguous(nt_input),
```
side note: the definition of `nested_tensor_impl_is_contiguous()` is expanded to handle strides in Yifan's PR #80981 (which was reverted but will be relanded).

AFAICT this PR won't need updates so just FYI :)

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -665,6 +665,80 @@ Tensor& NestedTensor_mul__Scalar(Tensor& self, const Scalar& other) {
+    OptionalIntArrayRef opt_dims,
+    bool keepdim,
+    c10::optional<ScalarType> dtype) {
+  // Only allow reductions across the ragged (last) dim
```
Note that NT supports raggedness in dims that are not the last dim

mikaylagawarecki replied:
Just to make sure I'm understanding this correctly, is it the case that only 1 dimension can be ragged but this dimension could be any dim?

jbschlosser replied:
NT also supports multiple ragged dims for any dim but the first (i.e. the implicit batch dimension). so you could have:

```python
torch.nested_tensor([torch.randn(2, 3), torch.randn(4, 5)])
```

where both dims 1 and 2 are ragged.

mikaylagawarecki replied:
Ah ok, that was what I thought! I only support the last dim because in the above example I don't think we would be able to sum across dim=1 unless we treat the tensors as being implicitly padded with 0s, whereas we can always sum across dim=-1. Does that sound right?

My bad, I think my comment here was misleading.

mikaylagawarecki replied:
And yes the test cases should be testing this case since for example 

`self.random_nt(device, dtype, 4, (4, 4))` would generate

```
nested_tensor([
  tensor([], size=(0, 0)),
  tensor([], size=(0, 1)),
  tensor([[-0.0821],
          [ 0.6607],
          [ 0.0785]]),
  tensor([[ 0.7884,  0.1021],
          [-0.2590, -0.1549]])
])
```

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -665,6 +665,80 @@ Tensor& NestedTensor_mul__Scalar(Tensor& self, const Scalar& other) {
+  // Always keep reduced dim for now
+  // This is to avoid the case where the nested tensors are 1D and keepdim=False
+  // making the nested tensors -> elements (e.g. sum(nt([1, 2 ,3], [4, 5]), -1) -> nt(6, 9))
+  TORCH_CHECK(keepdim, "NestedTensor always keeps the reduced dimension for now.");
```
nit: to make it a little bit clearer how to get around this error, maybe mention that `keepdim=True` is required

cpuhrsch commented:
An interesting follow up will be autograd, since it requires expand. We'll likely need to update the derivative formula. cc @drisspg

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -665,6 +665,80 @@ Tensor& NestedTensor_mul__Scalar(Tensor& self, const Scalar& other) {
 }
 
+// Very rudimentary sum_dim for prototyping with torch_scatter.segment_reduce.
+Tensor NestedTensor_sum__dim(
```
should this be named with CPU since it has CPU-specific kernel logic in it?

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -665,6 +665,80 @@ Tensor& NestedTensor_mul__Scalar(Tensor& self, const Scalar& other) {
+      int64_t segments = num_segments[i].item<int64_t>();
+      int64_t segment_length = segment_lengths[i].item<int64_t>();
+      for (const auto j : c10::irange(segments)) {
+        (void)j;
```
minor style nit: I think you wouldn't need this with a standard (non-range-based) for loop

jbschlosser reviewed (commented):
Thanks for the update! Looks good for exploration purposes :)

mikaylagawarecki committed:
Update on "Add rudimentary NestedTensor.sum(dim)"


A first step towards adding dimension-wise reductions to NestedTensor, 
- Assumes tensors in the nested tensor as well as the buffer of the nested tensor are contiguous
- Always enforces `keepdim=True`
- Only supports reduction across the last (ragged) dimension
- No support for acctype (`dtype` argument)
- No autograd support
- CPU only

Next steps would be to add support for the above. For now this basic support is for prototyping to make sure `NestedTensor` can be used as an API for segment reductions.



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add rudimentary NestedTensor.sum(dim)"


A first step towards adding dimension-wise reductions to NestedTensor, 
- Assumes tensors in the nested tensor as well as the buffer of the nested tensor are contiguous
- Always enforces `keepdim=True`
- Only supports reduction across the last (ragged) dimension
- No support for acctype (`dtype` argument)
- No autograd support
- CPU only

Next steps would be to add support for the above. For now this basic support is for prototyping to make sure `NestedTensor` can be used as an API for segment reductions.



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add rudimentary NestedTensor.sum(dim)"


A first step towards adding dimension-wise reductions to NestedTensor, 
- Assumes tensors in the nested tensor as well as the buffer of the nested tensor are contiguous
- Always enforces `keepdim=True`
- Only supports reduction across the last (ragged) dimension
- No support for acctype (`dtype` argument)
- No autograd support
- CPU only

Next steps would be to add support for the above. For now this basic support is for prototyping to make sure `NestedTensor` can be used as an API for segment reductions.



[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add rudimentary NestedTensor.sum(dim)"


A first step towards adding dimension-wise reductions to NestedTensor, 
- Assumes tensors in the nested tensor as well as the buffer of the nested tensor are contiguous
- Always enforces `keepdim=True`
- Only supports reduction across the last dimension
- No support for acctype (`dtype` argument)
- No autograd support
- CPU only

Next steps would be to add support for the above. For now this basic support is for prototyping to make sure `NestedTensor` can be used as an API for segment reductions.



[ghstack-poisoned]

jbschlosser reviewed (approved):
LGTM!

cpuhrsch commented:
cc @Acciorocketships

mikaylagawarecki commented:
@pytorchbot merge -g
